src/server/admin/models.py

```diff
@@ -35,61 +35,61 @@ def __init__(self, api_key: str, email: str = None) -> None:
 
     @staticmethod
     def list_users() -> List["User"]:
-        return session.query(User).all()
+        with Session() as session:
+            return session.query(User).all()
 
     @property
     def as_dict(self):
-        user_dict = deepcopy(self.__dict__)  # NOTE: changed from `self.__dict__.copy()` as it caused issues
-        # so we dont change the internal representation of self
-        user_dict["roles"] = self.get_user_roles
-        try:
-            return {k: user_dict[k] for k in ["id", "api_key", "email", "roles", "created", "last_time_used"]}
-        except KeyError:
-            return {
-                "id": self.id,
-                "api_key": self.api_key,
-                "email": self.email,
-                "roles": self.get_user_roles,
-                "created": self.created,
-                "last_time_used": self.last_time_used
-            }
+        return {
+            "id": self.id,
+            "api_key": self.api_key,
+            "email": self.email,
+            "roles": User.get_user_roles(self.id),
+            "created": self.created,
+            "last_time_used": self.last_time_used
+        }
 
-    @property
-    def get_user_roles(self) -> Set[str]:
-        return set([role.name for role in self.roles])
+    @staticmethod
+    def get_user_roles(user_id: int) -> Set[str]:
+        with Session() as session:
+            user = session.query(User).filter(User.id == user_id).first()
+            return set([role.name for role in user.roles])
 
     def has_role(self, required_role: str) -> bool:
-        return required_role in self.get_user_roles
+        return required_role in User.get_user_roles(self.id)
 
     @staticmethod
-    def assign_roles(user: "User", roles: Optional[Set[str]]) -> None:
+    def _assign_roles(user: "User", roles: Optional[Set[str]], session) -> None:
         get_structured_logger("api_user_models").info("setting roles", roles=roles, user_id=user.id, api_key=user.api_key)
         if roles:
+            db_user = session.query(User).filter(User.id == user.id).first() 
```

Comment:
should this be moved outside the if-statement? otherwise the else branch isn't going to have a `db_user` to assign roles to

Reply:
indeed!  good catch!